src/equations/shallow_water_two_layer_2d.jl

```diff
@@ -0,0 +1,936 @@
+``v_{2,lower}`` and ``v_{2,upper}``. The gravitational constant is denoted by `g`, the layer densitites by 
+``\rho_{upper}``and ``\rho_{lower}`` and the (possibly) variable bottom topography function by ``b(x)``. 
+Conservative variable water height ``h_{lower}`` is measured from the bottom topography ``b`` and ``h_{upper}`` 
+relative to ``h_{lower}``, therefore one also defines the total water heights as ``H_{lower} = h_{lower} + b`` and 
+``H_{upper} = h_{upper} + h_{lower} + b``.
+
+The densities must be chosen such that ``\rho_{upper} < \rho_{lower}``, to make sure that the heavier fluid 
+``\rho_{lower}`` is in the bottom layer and the lighter fluid ``\rho_{upper}`` in the upper layer.
+
+The additional quantity ``H_0`` is also available to store a reference value for the total water
+height that is useful to set initial conditions or test the "lake-at-rest" well-balancedness.
+
+The bottom topography function ``b(x)`` is set inside the initial condition routine
+for a particular problem setup.
+
+In addition to the unknowns, Trixi currently stores the bottom topography values at the 
+approximation points despite being fixed in time. This is done for convenience of computing the 
+bottom topography gradients on the fly during the approximation as well as computing auxiliary 
+quantities like the total water height ``H`` or the entropy variables.
+This affects the implementation and use of these equations in various ways:
+* The flux values corresponding to the bottom topography must be zero.
+* The bottom topography values must be included when defining initial conditions, boundary 
+  conditions or source terms.
+* [`AnalysisCallback`](@ref) analyzes this variable.
+* Trixi's visualization tools will visualize the bottom topography by default.
+
+A good introduction for the 2LSWE is available in Chapter 12 of the book:
+  - Benoit Cushman-Roisin (2011)\
+    Introduction to geophyiscal fluid dynamics: physical and numerical aspects\
+    <https://www.sciencedirect.com/bookseries/international-geophysics/vol/101/suppl/C>\
+    ISBN: 978-0-12-088759-0
+"""
+struct ShallowWaterTwoLayerEquations2D{RealT<:Real} <: AbstractShallowWaterEquations{2, 7}
+  gravity::RealT # gravitational constant
+  H0::RealT      # constant "lake-at-rest" total water height
+  rho_upper::RealT    # lower layer density
+  rho_lower::RealT    # upper layer density
+  r::RealT       # ratio of rho_upper / rho_lower
+end
+
+# Allow for flexibility to set the gravitational constant within an elixir depending on the
+# application where `gravity_constant=1.0` or `gravity_constant=9.81` are common values.
+# The reference total water height H0 defaults to 0.0 but is used for the "lake-at-rest"
+# well-balancedness test cases. Densities must be specificed such that rho_upper < rho_lower.
+function ShallowWaterTwoLayerEquations2D(; gravity_constant, H0=0.0, rho_upper, rho_lower)
```

Comment:
```suggestion
function ShallowWaterTwoLayerEquations2D(; gravity_constant, H0=zero(gravity_constant), rho_upper, rho_lower)
```
For type consistency (please confirm @ranocha)